Add capability to perform basic symbolic computations

nightscape opened the pull request:
This PR is _not yet ready to merge_ (it lacks tests, documentation and some improvements), but we'd like to get early feedback and code reviews.
It contains a set of SymbolicFunction classes and type classes to perform derivation, simplification, vectorization and conversion into ordinary Scala functions on them.
Currently, all computations are done by the compiler itself, although one could consider the use of macros which employ some JVM-based Computational Algebra System to perform advanced simplifications.

We're happy about all forms of feedback, especially about ideas on how to perform a function simplification that doesn't stop after one pass.

dlwh commented on math/src/main/scala/breeze/symbolic/CanBuildSingleArgumentFunction.scala:
```diff
@@ -0,0 +1,18 @@
+}
+
+object CanBuildSingleArgumentFunction {
+  implicit def canBuildExponential =
```
implicit object is probably better

dlwh commented on math/src/main/scala/breeze/symbolic/CanBuildSingleArgumentFunction.scala:
```diff
@@ -0,0 +1,18 @@
+      override def build[I <: SymbolicFunction](i: I): Exponential[I] = new Exponential(i)
+    }
+
+  implicit def canBuildLogarithm =
```
same

dlwh commented on math/src/main/scala/breeze/symbolic/CanCreateFunction.scala:
```diff
@@ -0,0 +1,167 @@
+trait LowPriorityCanCreateFunction {
+  implicit def canCreateVectorizedSymbolicFunction[I, O, F <: SymbolicFunction](
+    implicit canCreateFunction: CanCreateFunction[I, O, F],
+    bf: breeze.linalg.support.CanMapKeyValuePairs[DenseVector[I], Int, I, O, DenseVector[O]]
```
this is going to be highly inefficient in general. maybe that's fine?

nightscape replied:
This is the fallback if no other conversion options can be found. Do you have an idea on what we could improve in the general case?

dlwh commented on math/src/main/scala/breeze/symbolic/CanCreateFunction.scala:
```diff
@@ -0,0 +1,167 @@
+        (i: I) => exp(zero(i))
+    }
+
+  implicit def canCreateVarFunction[T] = new CanCreateFunction[T, T, Var] {
```
(in general, all implicits should have explicit return types)

dlwh commented on math/src/main/scala/breeze/symbolic/CanDerive.scala:
```diff
@@ -0,0 +1,130 @@
+      SingletonOf[T, t.type](t)
+  }
+  implicit class RichDerivable[F <: SymbolicFunction](f: F) {
+    def derivation(implicit canDerive: CanDerive[F]): canDerive.D = canDerive.derivation(f)
```
why not just put this method on SymbolicFunction?

nightscape replied:
You mean like this?
```
trait SymbolicFunction {
  def derivation(implicit canDerive: CanDerive[this.type]): canDerive.D = canDerive.derivation(this)
}
```
That would be an option. I just followed the pattern I saw in other type-class examples where the data classes (like SymbolicFunction) are relatively dumb and don't know about the type-class concerns (like derivation).

I was even considering not restricting the type-classes on SymbolicFunction at all. That would allow users to plugin any "description of a function", even if the implementation is not under their control (a third-party library), given they can provide a way to derive/simplify/functionize. I only put the SymbolicFunction restriction in to facilitate debugging type errors.
But I can definitely change this if you prefer.

dlwh commented on math/src/main/scala/breeze/symbolic/SymbolicFunction.scala:
```diff
@@ -0,0 +1,102 @@
+  * Not sealed because we want to allow users to
+  * create their own subclasses.
+  */
+trait SymbolicFunction
```
inheriting from NumericOps[SymbolicFunction] and using UFuncs for operator implementations seems like it would be good

dlwh commented on math/src/main/scala/breeze/symbolic/SymbolicFunction.scala:
```diff
@@ -0,0 +1,102 @@
+  */
+trait SymbolicFunction
+
+case class Var() extends SymbolicFunction {
```
case object?

dlwh replied:
how do you deal with multiple variables/partial derivatives/gradients?

nightscape replied:
Currently not at all. In our case we only have one free variable. One could give Var a parameter which specifies the variable, although this would probably have to be a singleton in order for the compiler to be able to incorporate the information about different variables into the calculations.
But that is currently beyond the scope of what we need / can invest time to do...

dlwh commented on math/src/main/scala/breeze/symbolic/SymbolicFunction.scala:
```diff
@@ -0,0 +1,102 @@
+  * allows us to retain compile-time information on zeroness
+  * which can be used to e.g. simplify functions.
+  */
+case class ConstZero() extends ConstBase {
```
case object?

dlwh commented on math/src/main/scala/breeze/symbolic/SymbolicFunction.scala:
```diff
@@ -0,0 +1,102 @@
+  * which can be used to e.g. simplify functions.
+  */
+case class ConstZero() extends ConstBase {
+  override def toString = "0.00"
```
why not just 0.0.toString?

nightscape replied:
I wanted to be able to distinguish it from a `Const[Double](0.0)` because that's important information when trying to find out why e.g. a formula doesn't get simplified.

dlwh commented on math/src/main/scala/breeze/symbolic/SymbolicFunction.scala:
```diff
@@ -0,0 +1,102 @@
+}
+
+case class ConstOne() extends ConstBase {
+  override def toString = "1.00"
```
similarly

dlwh commented on math/src/main/scala/breeze/symbolic/CanSimplify.scala:
```diff
@@ -0,0 +1,252 @@
+
+object CanSimplify extends HighPriorityCanSimplify {
+  implicit class RichSimplifiable[F](f: F) {
+    def simplify(implicit canSimplify: CanSimplify[F]): canSimplify.R =
```
i've been leaning towards UFuncs for this rather than CanX traits

nightscape replied:
I can give this a try. I've been only thinking of UFuncs as a way to abstract over arity, but it might as well work nicely here :+1:
Would you do this for all of vectorization, derivation, simplification and functionization?

nightscape commented:
Hi David,

I played around with the code a bit, made `SymbolicFunction` extend `NumericOps` and added `UFunc`s for derivation (`derive`), simplification (`simplify`), vectorization (`vectorize`) and functionization (`toFunction`). The `toFunction` `UFunc` is a little awkward to call, so I guess that the more common scenario will be to use the provided implicit wrapper like this
```scala
import CanCreateFunction._
Const(3.0).toFunction[Double, Double]
```

I also changed several `implicit def`s to `implicit val`s where no type parameter or implicit argument was needed. I went for `implicit def` + `implicit val` because it feels more consistent than `implicit def` + `implicit object` and [doesn't seem to have much of an impact otherwise](http://stackoverflow.com/questions/33317974/scala-implicit-object-vs-implicit-val).

Regarding tests: Would you go for run-time or compile-time tests? Does breeze already have compile-time tests from which I could borrow?

Also, it would be nice to make the compiler perform simplifications until it reaches a fix-point, but I have no idea how to do that... Currently you have to call `simplify` multiple times.

nightscape commented:
Hi @dlwh,

we've changed the approach rather radically: Instead of trying to make the Scala compiler do all the symbolic computation, we wrote macros that transform `SymbolicFunction` into an AST of the [Symja](https://bitbucket.org/axelclk/symja_android_library/overview) library, let Symja do the derivation / simplification and transform the result into a `CanDerive` / `CanSimplify` instance.
This gives us much better simplifications and could probably also be used to do symbolic computations other than deriving and simplifying.
Symja has a dual-license model, where the source code is LGPL, but the byte-code is Apache 2.
AFAIS this should be ok as we're only using a JAR dependency, but if you see a potential for problems, I can contact the author and ask for explicit permission / clarification of this use case.

I've also started adding some tests, so this is getting serious :smile: 
It would be great if you could have another look if this is heading in the right direction.
